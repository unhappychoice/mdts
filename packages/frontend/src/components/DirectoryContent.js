"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const material_1 = require("@mui/material");
const FileTreeContext_1 = require("../contexts/FileTreeContext");
const filterFileTree = (tree, targetPath) => {
    if (targetPath === '') {
        const result = [];
        tree.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    const findChildren = (currentTree, pathSegments, currentSegmentIndex) => {
        if (currentSegmentIndex === pathSegments.length) {
            return currentTree;
        }
        const segment = pathSegments[currentSegmentIndex];
        for (const item of currentTree) {
            if (typeof item !== 'string') {
                const key = Object.keys(item)[0];
                const itemSegments = key.split('/');
                if (itemSegments[itemSegments.length - 1] === segment) {
                    const children = item[key];
                    if (Array.isArray(children)) {
                        return findChildren(children, pathSegments, currentSegmentIndex + 1);
                    }
                }
            }
        }
        return null;
    };
    const pathSegments = targetPath.split('/').filter(s => s !== '');
    const children = findChildren(tree, pathSegments, 0);
    if (children) {
        const result = [];
        children.forEach(item => {
            if (typeof item === 'string') {
                result.push(item.split('/').pop() || item);
            }
            else {
                const key = Object.keys(item)[0];
                const newObject = {};
                newObject[key.split('/').pop() || key] = item[key];
                result.push(newObject);
            }
        });
        return result;
    }
    return [];
};
const DirectoryContent = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
    const { fileTree: fullFileTree, loading, error } = (0, FileTreeContext_1.useFileTreeContext)();
    const fileTree = filterFileTree(fullFileTree, selectedDirectoryPath);
    const pathSegments = selectedDirectoryPath
        ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
        : [];
    if (error)
        return (0, jsx_runtime_1.jsxs)("p", { children: ["Error: ", error] });
    const handleItemClick = (itemPath, isDirectory) => {
        if (isDirectory) {
            onDirectorySelect(itemPath);
        }
        else {
            onFileSelect(itemPath);
        }
    };
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: Object.assign({ flexGrow: 1, p: 4, bgcolor: 'background.paper' }, (contentMode === 'fixed' && {
            maxWidth: '800px',
            margin: '0 auto',
        })), children: [selectedDirectoryPath && ((0, jsx_runtime_1.jsx)(material_1.Breadcrumbs, { "aria-label": "breadcrumb", sx: { mb: 4 }, children: pathSegments.map((segment, index) => {
                    const isLast = index === pathSegments.length - 1;
                    const path = pathSegments.slice(0, index + 1).join('/');
                    return isLast ? ((0, jsx_runtime_1.jsx)(material_1.Typography, { color: "text.primary", children: segment }, path)) : ((0, jsx_runtime_1.jsx)(material_1.Link, { color: "inherit", href: "#", onClick: () => onDirectorySelect && onDirectorySelect(path), children: segment }, path));
                }) })), (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', mb: 8 }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { sx: { mr: 2 }, color: "primary", fontSize: "large" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h4", gutterBottom: true, sx: { mb: 0 }, children: selectedDirectoryPath })] }), (0, jsx_runtime_1.jsx)(material_1.Divider, { sx: { paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' } }), loading && ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })), (0, jsx_runtime_1.jsx)(material_1.Box, { mt: 4, children: fileTree && ((0, jsx_runtime_1.jsx)(material_1.List, { sx: { mr: -2, ml: -2 }, children: fileTree.map((item) => {
                        let name;
                        let itemPath;
                        let isDirectory;
                        if (typeof item === 'string') {
                            // It's a file
                            name = item;
                            itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                            isDirectory = false;
                        }
                        else {
                            // It's a directory object { [key: string]: FileTreeItem[] | string; }
                            const key = Object.keys(item)[0];
                            name = key;
                            itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                            isDirectory = true;
                        }
                        return ((0, jsx_runtime_1.jsxs)(material_1.ListItem, { button: true, onClick: () => handleItemClick(itemPath, isDirectory), children: [(0, jsx_runtime_1.jsx)(material_1.ListItemIcon, { sx: { minWidth: "38px" }, children: isDirectory ? (0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { color: "primary" }) : (0, jsx_runtime_1.jsx)(icons_material_1.ArticleOutlined, {}) }), (0, jsx_runtime_1.jsx)(material_1.ListItemText, { primary: name })] }, itemPath));
                    }) })) })] }));
};
exports.default = DirectoryContent;
