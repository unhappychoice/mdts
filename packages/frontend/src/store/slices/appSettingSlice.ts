import { createSlice } from '@reduxjs/toolkit';

type DarkMode = 'dark' | 'light' | 'auto';
type ContentMode = 'full' | 'compact';

interface AppSettingState {
  darkMode: DarkMode;
  contentMode: ContentMode;
  fileTreeOpen: boolean;
  outlineOpen: boolean;
}

const initialState: AppSettingState = {
  darkMode: 'auto',
  contentMode: 'compact',
  fileTreeOpen: true,
  outlineOpen: true,
};

const appSettingSlice = createSlice({
  name: 'appSetting',
  initialState,
  reducers: {
    saveAppSetting: (state, action: PayloadAction<{ darkMode: DarkMode, contentMode: ContentMode }>) => {
      state.darkMode = action.payload.darkMode;
      state.contentMode = action.payload.contentMode;
    },
    toggleFileTree: (state) => {
      state.fileTreeOpen = !state.fileTreeOpen;
    },
    toggleOutline: (state) => {
      state.outlineOpen = !state.outlineOpen;
    },
  },
});

export const { saveAppSetting, toggleFileTree, toggleOutline } = appSettingSlice.actions;

export default appSettingSlice.reducer;
