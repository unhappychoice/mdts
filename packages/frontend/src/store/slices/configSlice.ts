import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export const fetchConfig = createAsyncThunk('config/fetchConfig', async () => {
  return await fetchData<ConfigState>('/api/config', 'json');
});

export const saveConfigToBackend = createAsyncThunk(
  'config/saveConfigToBackend',
  async (config: ConfigState, { dispatch }) => {
    try {
      await fetch('/api/config', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
      });
      dispatch(fetchConfig()); // Re-fetch config after saving
    } catch (error) {
      console.error('Failed to save settings:', error);
      throw error; // Re-throw to allow error handling in component
    }
  }
);

export interface ConfigState {
  fontFamily: string;
  fontFamilyMonospace: string;
  fontSize: number;
  syntaxHighlighterTheme: string;
}

const initialState: ConfigState = {
  fontFamily: 'Roboto',
  fontFamilyMonospace: 'monospace',
  fontSize: 14,
  syntaxHighlighterTheme: 'auto',
};

const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action: PayloadAction<ConfigState>) => {
      state.fontFamily = action.payload.fontFamily;
      state.fontFamilyMonospace = action.payload.fontFamilyMonospace;
      state.fontSize = action.payload.fontSize;
      state.syntaxHighlighterTheme = action.payload.syntaxHighlighterTheme;
    });
  },
});

export default configSlice.reducer;
