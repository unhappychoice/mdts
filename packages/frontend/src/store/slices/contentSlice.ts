
import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface ContentState {
  content: string;
  loading: boolean;
  error: string | null;
  scrollPosition: number;
}

const initialState: ContentState = {
  content: '',
  loading: true,
  error: null,
  scrollPosition: 0,
};

export const fetchContent = createAsyncThunk(
  'content/fetchContent',
  async (path: string | null) => {
    const url = path ? `/api/markdown/${path}` : '/api/markdown/mdts-welcome-markdown.md';
    const data = await fetchData<string>(url, 'text');
    return data || '';
  }
);

const contentSlice = createSlice({
  name: 'content',
  initialState,
  reducers: {
    setScrollPosition: (state, action: PayloadAction<number>) => {
      state.scrollPosition = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchContent.pending, (state) => {
        if (!state.content) {
          state.loading = true;
        }
        state.error = null;
      })
      .addCase(fetchContent.fulfilled, (state, action) => {
        state.loading = false;
        state.content = action.payload;
      })
      .addCase(fetchContent.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch content';
      });
  },
});

export const { setScrollPosition } = contentSlice.actions;

export default contentSlice.reducer;
