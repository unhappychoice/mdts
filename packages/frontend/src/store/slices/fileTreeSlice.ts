
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface FileTreeItem {
  [key: string]: (FileTreeItem | string)[];
}

interface FileTreeState {
  fileTree: (FileTreeItem | string)[];
  filteredFileTree: (FileTreeItem | string)[];
  searchQuery: string;
  expandedNodes: string[];
  mountedDirectoryPath: string;
  loading: boolean;
  error: string | null;
}

const initialState: FileTreeState = {
  fileTree: [],
  filteredFileTree: [],
  searchQuery: '',
  expandedNodes: [],
  mountedDirectoryPath: '',
  loading: true,
  error: null,
};

const LOCAL_STORAGE_KEY_PREFIX = 'mdts_expanded_nodes_';
const LOCAL_STORAGE_RECENT_PATHS_KEY = 'mdts_recent_paths';
const MAX_RECENT_PATHS = 10;

const saveExpandedNodes = (path: string, nodes: string[]) => {
  try {
    localStorage.setItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`, JSON.stringify(nodes));
    let recentPaths: string[] = JSON.parse(localStorage.getItem(LOCAL_STORAGE_RECENT_PATHS_KEY) || '[]');
    recentPaths = recentPaths.filter(p => p !== path);
    recentPaths.unshift(path);
    if (recentPaths.length > MAX_RECENT_PATHS) {
      const oldPath = recentPaths.pop();
      if (oldPath) {
        localStorage.removeItem(`${LOCAL_STORAGE_KEY_PREFIX}${oldPath}`);
      }
    }
    localStorage.setItem(LOCAL_STORAGE_RECENT_PATHS_KEY, JSON.stringify(recentPaths));
  } catch (e) {
    console.error('Failed to save expanded nodes to local storage', e);
  }
};

const loadExpandedNodes = (path: string): string[] => {
  try {
    const storedNodes = localStorage.getItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`);
    return storedNodes ? JSON.parse(storedNodes) : [];
  } catch (e) {
    console.error('Failed to load expanded nodes from local storage', e);
    return [];
  }
};

const filterTree = (tree: (FileTreeItem | string)[], searchQuery: string): (FileTreeItem | string)[] => {
  if (!searchQuery) return tree;

  return tree
    .reduce((acc: (FileTreeItem | string)[], item) => {
      if (typeof item === 'string') {
        const fileName = item.split('/').pop() || '';
        return fileName.toLowerCase().includes(searchQuery.toLowerCase())
          ? [...acc, item]
          : acc;
      } else {
        const key = Object.keys(item)[0];
        const value = item[key];

        const children = Array.isArray(value)
          ? filterTree(value, searchQuery)
          : [];

        return children.length > 0
          ? [...acc, { [key]: children } as FileTreeItem ]
          :acc;
      }
    }, []);
};

export const fetchFileTree = createAsyncThunk(
  'fileTree/fetchFileTree',
  async () => {
    const data = await fetchData<{ fileTree: FileTreeItem[], mountedDirectoryPath: string }>('/api/filetree', 'json');
    return { fileTree: data?.fileTree || [], mountedDirectoryPath: data?.mountedDirectoryPath };
  }
);

const fileTreeSlice = createSlice({
  name: 'fileTree',
  initialState,
  reducers: {
    setSearchQuery: (state, action: { payload: string }) => {
      state.searchQuery = action.payload;
      state.filteredFileTree = filterTree(state.fileTree, action.payload);
    },
    toggleNode: (state, action: { payload: string }) => {
      const path = action.payload;
      if (state.expandedNodes.includes(path)) {
        state.expandedNodes = state.expandedNodes.filter((nodePath) => nodePath !== path);
      } else {
        state.expandedNodes.push(path);
      }
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setExpandedNodes: (state, action: { payload: string[] }) => {
      state.expandedNodes = action.payload;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    expandAllNodes: (state, action: { payload: (FileTreeItem | string)[] | null }) => {
      const allItemIds: string[] = [];
      const collectIds = (items: (FileTreeItem | string)[], parentPath: string = '') => {
        items.forEach(item => {
          if (typeof item !== 'string') {
            const key = Object.keys(item)[0];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            allItemIds.push(currentPath);
            collectIds(item[key] as (FileTreeItem | string)[], currentPath);
          }
        });
      };
      if (action.payload) {
        collectIds(action.payload);
      }
      state.expandedNodes = allItemIds;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setMountedDirectoryPath: (state, action: { payload: string }) => {
      state.mountedDirectoryPath = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchFileTree.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchFileTree.fulfilled, (state, action) => {
        state.loading = false;
        state.fileTree = action.payload.fileTree;
        state.mountedDirectoryPath = action.payload.mountedDirectoryPath;
        state.filteredFileTree = filterTree(action.payload.fileTree, state.searchQuery);
        state.expandedNodes = loadExpandedNodes(action.payload.mountedDirectoryPath);
      })
      .addCase(fetchFileTree.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch file tree';
      });
  },
});

export const selectFilteredFileTree = (
  fullFileTree: (FileTreeItem | string)[],
  targetPath: string
): (FileTreeItem | string)[] => {
  if (targetPath === '') {
    const result: FileTreeItem[] | string[] = [];
    fullFileTree.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  const findChildren = (
    currentTree: (FileTreeItem | string)[],
    pathSegments: string[],
    currentSegmentIndex: number
  ): (FileTreeItem | string)[] | null => {
    if (currentSegmentIndex === pathSegments.length) {
      return currentTree;
    }

    const segment = pathSegments[currentSegmentIndex];

    for (const item of currentTree) {
      if (typeof item !== 'string') {
        const key = Object.keys(item)[0];
        const itemSegments = key.split('/');
        if (itemSegments[itemSegments.length - 1] === segment) {
          const children = item[key];
          if (Array.isArray(children)) {
            return findChildren(children, pathSegments, currentSegmentIndex + 1);
          }
        }
      }
    }
    return null;
  };

  const pathSegments = targetPath.split('/').filter(s => s !== '');
  const children = findChildren(fullFileTree, pathSegments, 0);

  if (children) {
    const result: FileTreeItem[] | string[] = [];
    children.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  return [];
};

export const {
  setSearchQuery,
  toggleNode,
  setExpandedNodes,
  expandAllNodes,
  setMountedDirectoryPath,
} = fileTreeSlice.actions;

export default fileTreeSlice.reducer;
