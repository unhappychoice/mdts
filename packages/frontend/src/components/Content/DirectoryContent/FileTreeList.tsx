import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import { List, ListItem, ListItemIcon, ListItemText } from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { FileTreeItem } from '../../../store/slices/fileTreeSlice';
import { RootState } from '../../../store/store';

export interface FileTreeListProps {
  fileTree: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[];
  handleItemClick: (itemPath: string, isDirectory: boolean) => void;
}

export const FileTreeList: React.FC<FileTreeListProps> = ({ fileTree, handleItemClick }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);

  return (
    <List sx={{ mr: -2, ml: -2 }}>
      {fileTree.map((item) => {
        const isDirectory = !('path' in item);
        const name = isDirectory ? Object.keys(item)[0] : (item as FileTreeItem).path.split('/').pop();
        const itemPath = currentPath === '' ? name : `${currentPath}/${name}`;

        const handleClick = () => {
          handleItemClick(itemPath, isDirectory);
        };

        return (
          <ListItem key={itemPath} button={true} onClick={handleClick}>
            <ListItemIcon sx={{ minWidth: "38px" }}>
              {isDirectory ? <FolderOutlined color="primary" /> : <ArticleOutlined />}
            </ListItemIcon>
            <ListItemText primary={name} />
          </ListItem>
        );
      })}
    </List>
  );
};
