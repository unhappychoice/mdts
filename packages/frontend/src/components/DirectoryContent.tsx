import { ArticleOutlined, FolderOutlined } from '@mui/icons-material';
import {
  Box,
  Breadcrumbs,
  CircularProgress,
  Divider,
  Link,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography
} from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../store/store';
import { selectFilteredFileTree } from '../store/slices/fileTreeSlice';

interface DirectoryContentProps {
  selectedDirectoryPath: string;
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
  contentMode?: 'fixed' | 'full';
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ selectedDirectoryPath, onFileSelect, onDirectorySelect, contentMode = 'fixed' }) => {
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);
  const fileTree = selectFilteredFileTree(fullFileTree, selectedDirectoryPath);

  const pathSegments = selectedDirectoryPath
    ? selectedDirectoryPath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  const handleItemClick = (itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  };

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      {selectedDirectoryPath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 8 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large"/>
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {selectedDirectoryPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        {fileTree && (
          <List sx={{ mr: -2, ml: -2 }}>
            {fileTree.map((item) => {
              let name: string;
              let itemPath: string;
              let isDirectory: boolean;

              if (typeof item === 'string') {
                // It's a file
                name = item;
                itemPath = selectedDirectoryPath === '' ? item : `${selectedDirectoryPath}/${item}`;
                isDirectory = false;
              } else {
                // It's a directory object { [key: string]: FileTreeItem[] | string; }
                const key = Object.keys(item)[0];
                name = key;
                itemPath = selectedDirectoryPath === '' ? key : `${selectedDirectoryPath}/${key}`;
                isDirectory = true;
              }

              return (
                <ListItem button key={itemPath} onClick={() => handleItemClick(itemPath, isDirectory)}>
                  <ListItemIcon sx={{ minWidth: "38px" }}>
                    {isDirectory ? <FolderOutlined color="primary" /> : <ArticleOutlined />}
                  </ListItemIcon>
                  <ListItemText primary={name} />
                </ListItem>
              );
            })}
          </List>
        )}
      </Box>
    </Box>
  );
};

export default DirectoryContent;
