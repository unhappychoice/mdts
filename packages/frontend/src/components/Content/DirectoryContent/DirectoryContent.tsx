import { FolderOutlined } from '@mui/icons-material';
import { Box, CircularProgress, Divider, Typography } from '@mui/material';
import React, { useCallback } from 'react';
import { useSelector } from 'react-redux';
import { selectFilteredFileTree } from '../../../store/slices/fileTreeSlice';
import { RootState } from '../../../store/store';
import ErrorView from '../../ErrorView';
import BreadCrumb from '../BreadCrumb';
import { FileTreeList } from './FileTreeList';

interface DirectoryContentProps {
  onFileSelect: (filePath: string) => void;
  onDirectorySelect: (directoryPath: string) => void;
}

const DirectoryContent: React.FC<DirectoryContentProps> = ({ onFileSelect, onDirectorySelect }) => {
  const { currentPath } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);
  const { fileTree: fullFileTree, loading, error } = useSelector((state: RootState) => state.fileTree);

  const fileTree = selectFilteredFileTree(fullFileTree, currentPath);

  const handleItemClick = useCallback((itemPath: string, isDirectory: boolean) => {
    if (isDirectory) {
      onDirectorySelect(itemPath);
    } else {
      onFileSelect(itemPath);
    }
  }, [onDirectorySelect, onFileSelect]);

  if (error) {
    return <ErrorView error={error} />;
  }

  return (
    <Box
      sx={{
        width: '100%',
        minHeight: 'calc(100vh - 64px)',
        m: 0,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'compact' && {
          width: '800px',
          margin: '0 auto',
        })
      }}
    >
      <BreadCrumb onDirectorySelect={onDirectorySelect} />
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <FolderOutlined sx={{ mr: 2 }} color="primary" fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {currentPath}
        </Typography>
      </Box>
      <Divider sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }} />
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      )}
      <Box mt={4}>
        <FileTreeList fileTree={fileTree} handleItemClick={handleItemClick} />
      </Box>
    </Box>
  );
};

export default DirectoryContent;
