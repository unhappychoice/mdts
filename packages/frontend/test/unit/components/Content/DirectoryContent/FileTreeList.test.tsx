import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { FileTreeList } from '../../../../../src/components/Content/DirectoryContent/FileTreeList';
import { FileTreeItem } from '../../../../../src/store/slices/fileTreeSlice';

const mockStore = configureStore([]);

describe('FileTreeList', () => {
  let store;
  const mockFileTree: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[] = [
    { 'folder1': [] },
    { path: 'file2.txt', status: ' ' },
  ];

  beforeEach(() => {
    store = mockStore({
      history: {
        currentPath: '',
      },
    });
  });

  test('renders correctly with files and folders', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <FileTreeList fileTree={mockFileTree} handleItemClick={jest.fn()} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('folder1')).toBeInTheDocument();
    expect(screen.getByText('file2.txt')).toBeInTheDocument();
  });

  test('calls handleItemClick with correct arguments for a file', () => {
    const handleItemClickMock = jest.fn();
    render(
      <Provider store={store}>
        <FileTreeList fileTree={mockFileTree} handleItemClick={handleItemClickMock} />
      </Provider>
    );
    fireEvent.click(screen.getByText('file2.txt'));
    expect(handleItemClickMock).toHaveBeenCalledWith('file2.txt', false);
  });

  test('calls handleItemClick with correct arguments for a folder', () => {
    const handleItemClickMock = jest.fn();
    render(
      <Provider store={store}>
        <FileTreeList fileTree={mockFileTree} handleItemClick={handleItemClickMock} />
      </Provider>
    );
    fireEvent.click(screen.getByText('folder1'));
    expect(handleItemClickMock).toHaveBeenCalledWith('folder1', true);
  });

  test('handles currentPath correctly for item paths', () => {
    store = mockStore({
      history: {
        currentPath: 'parent_folder',
      },
    });
    const handleItemClickMock = jest.fn();
    render(
      <Provider store={store}>
        <FileTreeList fileTree={mockFileTree} handleItemClick={handleItemClickMock} />
      </Provider>
    );
    fireEvent.click(screen.getByText('file2.txt'));
    expect(handleItemClickMock).toHaveBeenCalledWith('parent_folder/file2.txt', false);
  });
});
