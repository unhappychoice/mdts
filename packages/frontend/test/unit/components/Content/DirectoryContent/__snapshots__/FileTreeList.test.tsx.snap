// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FileTreeList renders correctly with files and folders 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding css-1tfyysn-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-z2qibm-MuiListItem-root"
    >
      <div
        class="MuiListItemIcon-root css-1sms2q2-MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
          data-testid="FolderOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m9.17 6 2 2H20v10H4V6zM10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8z"
          />
        </svg>
      </div>
      <div
        class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1px98du-MuiTypography-root"
        >
          folder1
        </span>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-z2qibm-MuiListItem-root"
    >
      <div
        class="MuiListItemIcon-root css-1sms2q2-MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArticleOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
          />
          <path
            d="M14 17H7v-2h7zm3-4H7v-2h10zm0-4H7V7h10z"
          />
        </svg>
      </div>
      <div
        class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1px98du-MuiTypography-root"
        >
          file2.txt
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
